### ISSUE TITLE
Polkadot v0.9.5 sync error: storage root must match that calculated

### ISSUE THREAD

**Issue by gituser** [paritytech/polkadot#3327]
**Version:** `polkadot 0.9.5-3ce09da-x86_64-linux-gnu`
**OS:** Ubuntu Xenial (16.04)
**Polkadot cmdline:** `/home/polkadot/polkadot --name xxx --chain=westend -d /home/polkadot/.polkadot --rpc-external --ws-external --rpc-cors all --pruning=archive --disable-log-color --no-private-ipv4`

Just tried updating my `v0.9.3` polkadot binary to `v0.9.5` (built with `rustc v1.53.0`) on westend and that resulted in an error on syncing:

```
Jun 21 02:38:00 polkadot-testnet-dev polkadot[128]: 2021-06-21 02:38:00 ✨ Imported #6144499 (0x7b54…44b7)
Jun 21 02:38:02 polkadot-testnet-dev polkadot[128]: 2021-06-21 02:38:02 💤 Idle (25 peers), best: #6144499 (0x7b54…44b7), finalized #6144496 (0x3eff…4409), ⬇ 3.3kiB/s ⬆ 4.9kiB/s
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]: Hash: given=6470a17abde140dc9e77d0a4607096f09cbecc730a5bfee07f52c2af626d3abb, expected=8b937e351c240f4aa0643119488125a421dc13fdbd60c0a2f1c915ddc6f393e5
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]: ====================
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]: Version: 0.9.5-3ce09da-x86_64-linux-gnu
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    0: sp_panic_handler::set::{{closure}}
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    1: std::panicking::rust_panic_with_hook
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:              at rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:595:17
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    2: std::panicking::begin_panic::{{closure}}
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    3: std::sys_common::backtrace::__rust_end_short_backtrace
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    4: std::panicking::begin_panic
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    5: frame_executive::Executive<System,Block,Context,UnsignedValidator,AllPallets,COnRuntimeUpgrade>::final_checks
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    6: tracing::span::Span::in_scope
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    7: frame_executive::Executive<System,Block,Context,UnsignedValidator,AllPallets,COnRuntimeUpgrade>::execute_block
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    8: <westend_runtime::Runtime as sp_api::runtime_decl_for_Core::Core<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u32,sp_runtime::traits::BlakeTwo256>,sp_runtime::generic::unchecked_extrinsic::UncheckedExtrinsic<sp_runtime::multiaddress::MultiAddress<<<sp_runtime::MultiSignature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId,()>,westend_runtime::Call,sp_runtime::MultiSignature,(frame_system::extensions::check_spec_version::CheckSpecVersion<westend_runtime::Runtime>,frame_system::extensions::check_tx_version::CheckTxVersion<westend_runtime::Runtime>,frame_system::extensions::check_genesis::CheckGenesis<westend_runtime::Runtime>,frame_system::extensions::check_mortality::CheckMortality<westend_runtime::Runtime>,frame_system::extensions::check_nonce::CheckNonce<westend_runtime::Runtime>,frame_system::extensions::check_weight::CheckWeight<westend_runtime::Runtime>,pallet_transaction_payment::ChargeTransactionPayment<westend_runtime::Runtime>)>>>>::execute_block
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:    9: sp_api::runtime_decl_for_Core::execute_block_native_call_generator::{{closure}}
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   10: std::thread::local::LocalKey<T>::with
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   11: sc_executor::native_executor::WasmExecutor::with_instance::{{closure}}
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   12: sc_executor::wasm_runtime::RuntimeCache::with_instance
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   13: <sc_executor::native_executor::NativeExecutor<D> as sp_core::traits::CodeExecutor>::call
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   14: sp_state_machine::execution::StateMachine<B,H,N,Exec>::execute_aux
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   15: sp_state_machine::execution::StateMachine<B,H,N,Exec>::execute_using_consensus_failure_handler
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   16: <sc_service::client::call_executor::LocalCallExecutor<Block,B,E> as sc_client_api::call_executor::CallExecutor<Block>>::contextual_call
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   17: <sc_service::client::client::Client<B,E,Block,RA> as sp_api::CallApiAt<Block>>::call_api_at
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   18: sp_api::runtime_decl_for_Core::execute_block_call_api_at
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   19: <westend_runtime::RuntimeApiImpl<__SR_API_BLOCK__,RuntimeApiImplCall> as sp_api::Core<__SR_API_BLOCK__>>::Core_execute_block_runtime_api_impl
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   20: sp_api::Core::execute_block_with_context
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   21: sc_service::client::client::Client<B,E,Block,RA>::prepare_block_storage_changes
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   23: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   24: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   25: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   26: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   27: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   28: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   29: <sc_service::task_manager::prometheus_future::PrometheusFuture<T> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   30: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   31: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   32: <tracing_futures::Instrumented<T> as core::future::future::Future>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   33: std::thread::local::LocalKey<T>::with
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   34: futures_executor::local_pool::block_on
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   35: tokio::runtime::task::core::Core<T,S>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   36: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   37: tokio::runtime::task::harness::Harness<T,S>::poll
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   38: tokio::runtime::blocking::pool::Inner::run
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   39: tokio::runtime::context::enter
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   40: std::sys_common::backtrace::__rust_begin_short_backtrace
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   41: core::ops::function::FnOnce::call_once{{vtable.shim}}
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   42: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:              at rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/alloc/src/boxed.rs:1546:9
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:       <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:              at rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/alloc/src/boxed.rs:1546:9
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:       std::sys::unix::thread::Thread::new::thread_start
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:              at rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/sys/unix/thread.rs:71:17
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   43: start_thread
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]:   44: clone
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]: Thread 'tokio-runtime-worker' panicked at 'Storage root must match that calculated.', /home/build/.cargo/git/checkouts/substrate-7e08433d4c370a21/9c57262/frame/executive/src/lib.rs:462
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]: This is a bug. Please report it at:
Jun 21 02:38:06 polkadot-testnet-dev polkadot[128]: #011https://github.com/paritytech/polkadot/issues/new
```

Rolling back to v0.9.3 helps to prevent these errors.

**Comment by cy92830** [comment]
same issue found

**Comment by boohyung** [comment]
same issue is found.

polkadot command: `./target/release/polkadot --chain=westend --pruning=archive --rpc-external --ws-external --rpc-cors=all -l info -d ~/.dot`

error log:
```
Thread 'tokio-runtime-worker' panicked at 'Storage root must match that calculated.', /home/polkadot/.cargo/git/checkouts/substrate-7e08433d4c370a21/9c57262/frame/executive/src/lib.rs:462

This is a bug. Please report it at:

        https://github.com/paritytech/polkadot/issues/new
```

**Comment by pkrasam** [comment]
do we have a resolution for this error after the upgrade to `v0.9.5`
```
Jun 21 22:19:06 polkadot[932976]:    9: sp_api::runtime_decl_for_Core::execute_block_native_call_generator::{{closure}}
Jun 21 22:19:06 polkadot[932976]:   10: std::thread::local::LocalKey<T>::with
Jun 21 22:19:06 polkadot[932976]:   11: sc_executor::native_executor::WasmExecutor::with_instance::{{closure}}
Jun 21 22:19:06 polkadot[932976]:   12: sc_executor::wasm_runtime::RuntimeCache::with_instance
Jun 21 22:19:06 polkadot[932976]:   13: <sc_executor::native_executor::NativeExecutor<D> as sp_core::traits::CodeExecutor>::call
Jun 21 22:19:06 polkadot[932976]:   14: sp_state_machine::execution::StateMachine<B,H,N,Exec>::execute_aux
Jun 21 22:19:06 polkadot[932976]:   15: sp_state_machine::execution::StateMachine<B,H,N,Exec>::execute_using_consensus_failure_handler
Jun 21 22:19:06 polkadot[932976]:   16: <sc_service::client::call_executor::LocalCallExecutor<Block,B,E> as sc_client_api::call_executor::CallExecutor<Block>>::contextual_call
Jun 21 22:19:06 polkadot[932976]:   17: <sc_service::client::client::Client<B,E,Block,RA> as sp_api::CallApiAt<Block>>::call_api_at
Jun 21 22:19:06 polkadot[932976]:   18: sp_api::runtime_decl_for_Core::execute_block_call_api_at
Jun 21 22:19:06 polkadot[932976]:   19: <westend_runtime::RuntimeApiImpl<__SR_API_BLOCK__,RuntimeApiImplCall> as sp_api::Core<__SR_API_BLOCK__>>::Core_execute_block_runtime_api_impl
Jun 21 22:19:06 polkadot[932976]:   20: sp_api::Core::execute_block_with_context
Jun 21 22:19:06 polkadot[932976]:   21: sc_service::client::client::Client<B,E,Block,RA>::prepare_block_storage_changes
Jun 21 22:19:06 polkadot[932976]:   22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   23: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   24: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   25: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   26: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   27: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   28: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   29: <sc_service::task_manager::prometheus_future::PrometheusFuture<T> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   30: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   31: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   32: <tracing_futures::Instrumented<T> as core::future::future::Future>::poll
Jun 21 22:19:06 polkadot[932976]:   33: std::thread::local::LocalKey<T>::with
Jun 21 22:19:06 polkadot[932976]:   34: futures_executor::local_pool::block_on
Jun 21 22:19:06 polkadot[932976]:   35: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
Jun 21 22:19:06 polkadot[932976]:   36: tokio::runtime::task::raw::poll
Jun 21 22:19:06 polkadot[932976]:   37: tokio::runtime::blocking::pool::Inner::run
Jun 21 22:19:06 polkadot[932976]:   38: tokio::runtime::context::enter
Jun 21 22:19:06 polkadot[932976]:   39: std::sys_common::backtrace::__rust_begin_short_backtrace
Jun 21 22:19:06 polkadot[932976]:   40: core::ops::function::FnOnce::call_once{{vtable.shim}}
Jun 21 22:19:06 polkadot[932976]:   41: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
Jun 21 22:19:06 polkadot[932976]:              at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/alloc/src/boxed.rs:1546:9
Jun 21 22:19:06 polkadot[932976]:       <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
Jun 21 22:19:06 polkadot[932976]:              at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/alloc/src/boxed.rs:1546:9
Jun 21 22:19:06 polkadot[932976]:       std::sys::unix::thread::Thread::new::thread_start
Jun 21 22:19:06 polkadot[932976]:              at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/sys/unix/thread.rs:71:17
Jun 21 22:19:06 polkadot[932976]:   42: start_thread
Jun 21 22:19:06 polkadot[932976]:   43: clone
Jun 21 22:19:06 polkadot[932976]: Thread 'tokio-runtime-worker' panicked at 'Storage root must match that calculated.', /home/westend/.cargo/git/checkouts/substrate-7e08433d4c370a21/9c57262/frame/executive/src/lib.rs:462
Jun 21 22:19:06 polkadot[932976]: This is a bug. Please report it at:
Jun 21 22:19:06 polkadot[932976]:         https://github.com/paritytech/polkadot/issues/new
Jun 21 22:19:06 polkadot[932976]: 2021-06-21 22:19:06 ✨ Imported #6157895 (0x0456…a4c2)
```

**Comment by NukeManDan** [comment]
If I understand correctly, https://github.com/paritytech/polkadot/releases/tag/untagged-9af37a2701717eb6b716 includes a fix for this, thus 0.9.6 when released should be tried. Let me know if you still have issues with 0.9.6, be on the lookout for this in the next day or so.

**Comment by gituser** [comment]
In polkadot `v0.9.6` the issue seems to be fixed.
